File: ATTENDANCE/api/[...path].py
"""
Lightweight API proxy for Vercel.
Forwards /api/* requests to the Python backend (deployed on Render/Railway).
Set BACKEND_URL in Vercel environment variables.
"""
from http.server import BaseHTTPRequestHandler
import os
import urllib.request
import urllib.error

BACKEND_URL = os.environ.get("BACKEND_URL", "").rstrip("/")
SKIP_HEADERS = {"host", "connection", "transfer-encoding", "content-length"}


class handler(BaseHTTPRequestHandler):
    def _send_cors(self):
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type, Authorization")

    def _proxy(self, method):
        if not BACKEND_URL:
            self.send_response(503)
            self.send_header("Content-Type", "application/json")
            self._send_cors()
            self.end_headers()
            self.wfile.write(
                b'{"error":"BACKEND_URL is not set. Deploy the backend on Render and add BACKEND_URL in Vercel settings."}'
            )
            return

        content_length = int(self.headers.get("Content-Length", 0))
        body = self.rfile.read(content_length) if content_length else None

        url = f"{BACKEND_URL}{self.path}"
        headers = {}
        content_type = self.headers.get("Content-Type")
        if content_type:
            headers["Content-Type"] = content_type

        req = urllib.request.Request(url, data=body, headers=headers, method=method)

        try:
            with urllib.request.urlopen(req, timeout=55) as resp:
                self.send_response(resp.status)
                for key, value in resp.headers.items():
                    if key.lower() not in SKIP_HEADERS:
                        self.send_header(key, value)
                self._send_cors()
                self.end_headers()
                self.wfile.write(resp.read())
        except urllib.error.HTTPError as e:
            self.send_response(e.code)
            for key, value in e.headers.items():
                if key.lower() not in SKIP_HEADERS:
                    self.send_header(key, value)
            self._send_cors()
            self.end_headers()
            self.wfile.write(e.read())
        except Exception as e:
            self.send_response(502)
            self.send_header("Content-Type", "application/json")
            self._send_cors()
            self.end_headers()
            self.wfile.write(
                f'{{"error":"Backend unavailable. Check BACKEND_URL and ensure the backend is running. Details: {str(e)}"}}'.encode()
            )

    def do_GET(self):
        self._proxy("GET")

    def do_POST(self):
        self._proxy("POST")

    def do_PUT(self):
        self._proxy("PUT")

    def do_DELETE(self):
        self._proxy("DELETE")

    def do_OPTIONS(self):
        self.send_response(200)
        self._send_cors()
        self.end_headers()

File: api/[...path].js
/**
 * Vercel Serverless Function: API Proxy
 * Forwards /api/* requests to your Python backend (e.g. deployed on Render, Railway, etc.)
 * 
 * Set BACKEND_URL in your Vercel Project Settings -> Environment Variables.
 * Example: BACKEND_URL = https://attendance-system.onrender.com
 */

export const config = {
  api: {
    bodyParser: false, // Disabling bodyParser allows raw stream forwarding (essential for file uploads and downloads)
    responseLimit: false,
  },
};

export default async function handler(req, res) {
  // Always handle CORS preflight requests first
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS, PATCH');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization, X-Requested-With, Accept');

  if (req.method === 'OPTIONS') {
    return res.status(204).end();
  }

  const backendUrl = process.env.BACKEND_URL ? process.env.BACKEND_URL.replace(/\/+$/, '') : null;

  if (!backendUrl) {
    return res.status(503).json({
      error: 'BACKEND_URL is not set on Vercel.',
      message: 'Please set BACKEND_URL in your Vercel project Settings -> Environment Variables (e.g. https://your-backend.onrender.com). Alternatively, open the Settings modal in the web app to connect directly to your backend URL.',
      configured: false
    });
  }

  // Construct target URL
  // req.url contains full path e.g. /api/students or /api/reports/monthly?month=2026-09
  const targetPath = req.url.startsWith('/') ? req.url : `/${req.url}`;
  const targetUrl = `${backendUrl}${targetPath}`;

  try {
    // Read request body if present (POST, PUT, PATCH, DELETE)
    let bodyData = null;
    if (['POST', 'PUT', 'PATCH', 'DELETE'].includes(req.method)) {
      const chunks = [];
      for await (const chunk of req) {
        chunks.push(chunk);
      }
      if (chunks.length > 0) {
        bodyData = Buffer.concat(chunks);
      }
    }

    // Forward relevant headers
    const forwardHeaders = {};
    const skipHeaders = ['host', 'connection', 'content-length', 'transfer-encoding'];
    for (const [key, value] of Object.entries(req.headers)) {
      if (!skipHeaders.includes(key.toLowerCase())) {
        forwardHeaders[key] = value;
      }
    }

    // Set request timeout to 60s (handles Render cold-starts)
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), 60000);

    const backendResponse = await fetch(targetUrl, {
      method: req.method,
      headers: forwardHeaders,
      body: bodyData,
      signal: controller.signal,
      redirect: 'follow',
    });

    clearTimeout(timeoutId);

    // Forward response headers (excluding hop-by-hop headers and duplicate CORS)
    const skipResponseHeaders = ['transfer-encoding', 'connection', 'access-control-allow-origin'];
    backendResponse.headers.forEach((val, key) => {
      if (!skipResponseHeaders.includes(key.toLowerCase())) {
        res.setHeader(key, val);
      }
    });

    // Ensure single CORS header
    res.setHeader('Access-Control-Allow-Origin', '*');

    // Get response body as buffer and send with proper status
    const arrayBuffer = await backendResponse.arrayBuffer();
    const buffer = Buffer.from(arrayBuffer);

    res.status(backendResponse.status).send(buffer);

  } catch (error) {
    console.error('API Proxy Error:', error);

    const isTimeout = error.name === 'AbortError';
    const statusCode = isTimeout ? 504 : 502;

    return res.status(statusCode).json({
      error: isTimeout ? 'Gateway Timeout' : 'Bad Gateway',
      message: isTimeout
        ? 'The backend server took longer than 60s to respond. If running on Render free tier, it may still be waking up. Please retry in a few seconds.'
        : `Could not reach backend server at ${backendUrl}. Ensure the backend is active and reachable. Error: ${error.message}`,
      targetUrl,
      isWakeUpCandidate: true
    });
  }
}
